src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -3214,10 +3215,32 @@ internal static BuildIntegratedInstallationContext CreateInstallationContextForP
                 .Except(unsuccessfulFrameworks)
                 .ToList();
 
+            var areAllPackagesConditional = DoesPackageAppearWithDifferentVersions(packageIdentityId, originalPackageSpec);
+
             return new BuildIntegratedInstallationContext(
                 successfulFrameworksWithPackage,
                 unsuccessfulFrameworks.Union(frameworksWithoutResultingPackage).Distinct(),
-                originalFrameworks);
+                originalFrameworks,
+                areAllPackagesConditional);
+        }
+
+        private static bool DoesPackageAppearWithDifferentVersions(string packageIdentityId, PackageSpec packageSpec)
+        {
+            HashSet<VersionRange> versions = default;
+
+            foreach (var framework in packageSpec.TargetFrameworks)
```

Comment:
I am wondering if we should check for only `successfulFrameworksWithPackage` instead of all the target frameworks. The reason is we invoke conditional service only for successful frameworks [here](https://github.com/NuGet/NuGet.Client/pull/5532/files#diff-5fae96e8312674fe53149dbdb5a5f036aea82698a5aeff49617e8e7ae66d2685R387).

Reply:
This is a before thing, if you look at the calls, it's only using the original package spec, not the resulting spec.

In general though, my focus is on having the correct heuristic even if we're not short circuiting as early. 
These code paths are not very hot, and as such there's no need for a heavy performance focus at the cost of readability.